src/util/createInstantSearch.js

```diff
@@ -6,50 +6,109 @@ import { warn } from './warn';
       })
     );
 
-    return search.helper.searchOnce().then(({ content: lastResults }) => {
-      // The search instance needs to act as if this was a regular `search`
-      // but return a promise, since that is the interface of `asyncData`
-      search.helper.lastResults = lastResults;
+    derivedHelpers = Object.keys(uiStateMapping).reduce((acc, indexId) => {
+      // eslint-disable-next-line no-param-reassign
+      acc[indexId] = helper.derive(
+        () =>
+          new algoliaHelper.SearchParameters(
+            _objectSpread({}, uiStateMapping[indexId], {
+              // parameters set by default
+              highlightPreTag: '__ais-highlight__',
+              highlightPostTag: '__/ais-highlight__',
+            })
+          )
+      );
+      return acc;
+    }, {});
+
+    helper.searchOnlyWithDerivedHelpers();
+
+    return new Promise((resolve, reject) => {
+      // possible bug if something changes on the helper:
+      // we only wait for the result of the main derived helper,
+      // assuming _all_ results happen in the same tick
+      derivedHelpers[indexName].on('error', reject);
+      derivedHelpers[indexName].on('result', resolve);
+    }).then(() => {
+      // TODO: do we need this for anything else than our own warning?
+      search.helper = helper;
+      search.mainHelper = search.helper;
+      // set so root widgets, which don't have an ais-index, can render
+      search.helper.lastResults = derivedHelpers[indexName].lastResults;
+
+      // init main index here, so force render later works as expected
+      // because main index isn't a vue widget and doesn't get __forceRender
+      search.mainIndex.init({
+        instantSearchInstance: search,
+        parent: null,
+        uiState: initialUiState,
+      });
+      // force set last results on derived helper, since we search before init
+      // TODO: PR for hydrate on index widget
+      search.mainIndex.hydrate(derivedHelpers[indexName].lastResults);
     });
   };
 
   // make sure correct data is available in each widget's state
-  // called in widget mixin
-  search.__forceRender = widget => {
+  // called in widget mixin with (this.widget, this)
+  search.__forceRender = (widget, componentInstance) => {
     if (!search.helper) {
       warn(
         '`instantsearch.findResultsState()` needs to be called when using `ais-instant-search-ssr`.'
       );
       return;
     }
 
+    const parentIndex = componentInstance.getParentIndex();
+    const rootIndex = componentInstance.instantSearchInstance;
+
+    const parent = parentIndex !== rootIndex ? parentIndex : null;
+    const uiState = initialUiState;
+
+    // API to get helper & state is different for InstantSearch as Index widget
+    // so we need to call those conditionally only if parentIndex is the widget
+    const helper = parent ? parent.getHelper() : search.helper;
+    const results = parent ? parent.getResults() : search.helper.lastResults;
```

Comment:
could InstantSearch have these to be synchronous with an index widget? Or could we say `parent || search.mainIndex`?

Reply:
mainIndex as the fallback is the better solution